### PR TITLE
[WIP] Bring in build tools to CLI 

### PR DESCRIPTION
**Pull request by sokket**
Update the build script to pull down build tools into the build

/fyi: @piotrpMSFT @eerhardt @livarcocc @brthor

### CODE REVIEW

**Comment by sokket** [comment]
NOTE - non-windows is still a WIP

File: init-tools.cmd [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+set /P BUILDTOOLS_VERSION=< "%~dp0BuildToolsVersion.txt"
+set BUILD_TOOLS_PATH=%PACKAGES_DIR%Microsoft.DotNet.BuildTools\%BUILDTOOLS_VERSION%\lib\
+set PROJECT_JSON_PATH=%TOOLRUNTIME_DIR%\%BUILDTOOLS_VERSION%
+set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.json
+set PROJECT_JSON_CONTENTS={ "dependencies": { "Microsoft.DotNet.BuildTools": "%BUILDTOOLS_VERSION%" }, "frameworks": { "netcoreapp1.0": { } } }
+set BUILD_TOOLS_SEMAPHORE=%PROJECT_JSON_PATH%\init-tools.completed
+set DOTNET_INSTALL_DIR=%~dp0.dotnet_stage0
+
+if DEFINED "%CHANNEL%" (
+  set SCRIPT="%~dp0scripts\obtain\dotnet-install.ps1 -Channel %CHANNEL% -Architecture %bit% -InstallDir \"%DOTNET_PATH%\""
```
Comment (brthor) [inline init-tools.cmd]: Should this be using `DOTNET_INSTALL_DIR` instead of `DOTNET_PATH` ?

File: init-tools.cmd [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+if [%BUILDTOOLS_SOURCE%]==[] set BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
+set /P BUILDTOOLS_VERSION=< "%~dp0BuildToolsVersion.txt"
+set BUILD_TOOLS_PATH=%PACKAGES_DIR%Microsoft.DotNet.BuildTools\%BUILDTOOLS_VERSION%\lib\
+set PROJECT_JSON_PATH=%TOOLRUNTIME_DIR%\%BUILDTOOLS_VERSION%
+set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.json
+set PROJECT_JSON_CONTENTS={ "dependencies": { "Microsoft.DotNet.BuildTools": "%BUILDTOOLS_VERSION%" }, "frameworks": { "netcoreapp1.0": { } } }
+set BUILD_TOOLS_SEMAPHORE=%PROJECT_JSON_PATH%\init-tools.completed
+set DOTNET_INSTALL_DIR=%~dp0.dotnet_stage0
+
+if DEFINED "%CHANNEL%" (
```
Comment (eerhardt) [inline init-tools.cmd]: Where do we expect %CHANNEL% to be set?  Shouldn't it be read from the branchinfo.txt file?

File: init-tools.cmd [hunk 3]
```diff
@@ -0,0 +1,79 @@
+@if "%_echo%" neq "on" echo off
```
Comment (eerhardt) [inline init-tools.cmd]: Would it be better if this whole script was written in PS1?

File: build_projects/dotnet-cli-build/build.ps1 [hunk 4]
```diff
@@ -47,27 +47,27 @@ cat "$RepoRoot\branchinfo.txt" | ForEach-Object {
 }
 
 # Use a repo-local install directory (but not the artifacts directory because that gets cleaned a lot
-if (!$env:DOTNET_INSTALL_DIR)
-{
-    $env:DOTNET_INSTALL_DIR="$RepoRoot\.dotnet_stage0\Windows\$Architecture"
-}
+#if (!$env:DOTNET_INSTALL_DIR)
```
Comment (eerhardt) [inline build_projects/dotnet-cli-build/build.ps1]: I assume in the finished version, all this commented out code gets deleted.

File: init-tools.cmd [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+setlocal
+
+:CheckOS
+IF "%PROCESSOR_ARCHITECTURE%"=="x86" (set bit=x86) else (set bit=x64)
```
Comment (eerhardt) [inline init-tools.cmd]: The Architecture is defined in the parameters to "build.cmd".  See [https://github.com/dotnet/cli/blob/rel/1.0.0/netci.groovy#L29](https://github.com/dotnet/cli/blob/rel/1.0.0/netci.groovy#L29) for an example (we use the same approach in VSO). This needed to read that parameter.

File: init-tools.cmd [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+set INIT_TOOLS_LOG=%~dp0init-tools.log
+if [%PACKAGES_DIR%]==[] set PACKAGES_DIR=%~dp0packages\
```
Comment (eerhardt) [inline init-tools.cmd]: The CLI's nuget package is set to `.nuget`.  This should use the same.  We only need 1 package cache.

File: init-tools.cmd [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  if NOT exist "%DOTNET_INSTALL_DIR%" mkdir "%DOTNET_INSTALL_DIR%"
+  if NOT exist "%PACKAGES_DIR%" mkdir "%PACKAGES_DIR%"
+
+  echo %SCRIPT%
+  powershell -NoProfile -NoLogo -Command %SCRIPT% "-Verbose; exit $LastExitCode;"
+  if %errorlevel% neq 0 exit /b %errorlevel%
+  setx DOTNET_SKIP_FIRST_TIME_EXPERIENCE 1
+)
+
+:afterdotnetrestore
```
Comment (eerhardt) [inline init-tools.cmd]: Do we need this label?  it doesn't appear to be used.

File: init-tools.cmd (hunk 1 above)
Comment (brthor) [inline init-tools.cmd]: Or what is the difference between the 2?

File: init-tools.cmd (hunk 2 above)
Comment (sokket) [inline init-tools.cmd]: The PS1 script expected CHANNEL to be an environment variable, so this is the same

File: init-tools.cmd (hunk 3 above)
Comment (sokket) [inline init-tools.cmd]: I tried to keep the init-tools script similar to CoreFX for parity

File: build_projects/dotnet-cli-build/build.ps1 (hunk 4 above)
Comment (sokket) [inline build_projects/dotnet-cli-build/build.ps1]: Yup :)

File: init-tools.cmd (hunk 1 above)
Comment (sokket) [inline init-tools.cmd]: Renamed to `DOTNET_INSTALL_ROOT_DIR` to be more concise. The root needs to be created and then we simply extract there...the subfolders are an implementation detail

File: init-tools.cmd (hunk 6 above)
Comment (sokket) [inline init-tools.cmd]: Fixed

File: init-tools.cmd (hunk 7 above)
Comment (sokket) [inline init-tools.cmd]: Removed

File: BuildToolsVersion.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+1.0.26-prerelease-00524-01
```
Comment (piotrpMSFT) [inline BuildToolsVersion.txt]: @eerhardt is this something maestro can manage for us?

File: init-tools.cmd (hunk 5 above)
Comment (sokket) [inline init-tools.cmd]: Fixed

File: init-tools.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+__scriptpath=$(cd "$(dirname "$0")"; pwd -P)
```
Comment (brthor) [inline init-tools.sh]: nit: all the assignments should be quoted when set to a path

File: init-tools.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    if [ ! -e $__BUILD_TOOLS_PATH ]; then
+        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
+        echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE" >> $__init_tools_log
+        $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE >> $__init_tools_log
+        if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details."; fi
+    fi
+
+    echo "Initializing BuildTools..."
+    echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> $__init_tools_log
+    $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR >> $__init_tools_log
```
Comment (brthor) [inline init-tools.sh]: nit: and quoted as well when the parameters are used, to support spaces not breaking things

File: init-tools.cmd (hunk 5 above)
Comment (sokket) [inline init-tools.cmd]: Actually, after talking around I don't think we need the architecture passed around at all. When we pull down the binaries, we should just check what the current running architecture is and pull that one down. You can switch between 64 bit and 32 bit by having the right command prompt

File: build.cmd [hunk 11]
```diff
@@ -3,5 +3,15 @@
 REM Copyright (c) .NET Foundation and contributors. All rights reserved.
 REM Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+REM The Architecture parameter is the 4th item in the space delimited list, so check for it
```
Comment (eerhardt) [inline build.cmd]: I don't think it was the intention of that the Architecture was specified at the 4th item.  Instead "-Architecture value" should be able to be specified in any order.

File: BuildToolsVersion.txt (hunk 8 above)
Comment (eerhardt) [inline BuildToolsVersion.txt]: Yes, if 
1. build-tools starts writing out a build-info file to dotnet/versions.
2. We teach update-dependencies to look for that build-info file, and update this file as necessary.

---
In reply to: [68859625](https://github.com/dotnet/cli/pull/3719#discussion_r68859625) [](ancestors = 68859625)

File: init-tools.cmd (hunk 5 above)
Comment (eerhardt) [inline init-tools.cmd]: 1. that is a change in our current dev workflow.  I really liked to be able to just say "build -Architecture x86" and it just work from any of my command prompts.
2. Are we sure we can "have the right command prompt" on Jenkins and in VSO?

Let's not try to change this now.  Stick to what is currently working, and we can refactor later, if we feel necessary.

---
In reply to: [68865996](https://github.com/dotnet/cli/pull/3719#discussion_r68865996) [](ancestors = 68865996)

File: init-tools.cmd (hunk 2 above)
Comment (eerhardt) [inline init-tools.cmd]: The reason the PS1 script expected CHANNEL to be an env var, is because that PS1 script actually set the CHANNEL env var.  See [https://github.com/dotnet/cli/blob/rel/1.0.0/build_projects/dotnet-cli-build/build.ps1#L54-L60](https://github.com/dotnet/cli/blob/rel/1.0.0/build_projects/dotnet-cli-build/build.ps1#L54-L60).  This was setting CHANNEL to the value set in the branchinfo.txt file.

---
In reply to: [68847256](https://github.com/dotnet/cli/pull/3719#discussion_r68847256) [](ancestors = 68847256)

File: init-tools.cmd (hunk 5 above)
Comment (sokket) [inline init-tools.cmd]: But why do you even need to worry about if your stage0 is 64 or 32 bit?

File: init-tools.cmd (hunk 3 above)
Comment (eerhardt) [inline init-tools.cmd]: Is that a good enough reason?

1. This file is duplicated across repos, and trying to keep it "in sync" isn't going to be possible.
2. We are already changing this file, and in some cases pretty significantly (using dotnet-install, using the CHANNEL from the branchinfo.txt file, ).

---
In reply to: [68847295](https://github.com/dotnet/cli/pull/3719#discussion_r68847295) [](ancestors = 68847295)

File: build.sh [hunk 12]
```diff
@@ -52,6 +52,9 @@ dockerbuild()
     BUILD_COMMAND=/opt/code/build_projects/dotnet-cli-build/build.sh $DIR/scripts/dockerrun.sh --non-interactive "$@"
 }
 
+# Initialize the tooling
+$DIR/init-tools.sh "${args[@]}"
```
Comment (eerhardt) [inline build.sh]: Does this need to be run inside the docker container?  That way we get the correct stage0 for the actual platform we will be building on.

File: init-tools.sh [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+then
+    echo "Increasing file description limit to 1024"
+    ulimit -n 1024
+fi
+
+# Disable first run since we want to control all package sources
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+if [ ! -e $__PROJECT_JSON_FILE ]; then
+    $__scriptpath/scripts/obtain/dotnet-install.sh --install-dir $__DOTNET_PATH
```
Comment (eerhardt) [inline init-tools.sh]: This should use the CHANNEL variable set in branchinfo.txt.

File: init-tools.sh [hunk 14]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env bash
+
+__scriptpath=$(cd "$(dirname "$0")"; pwd -P)
+__init_tools_log=$__scriptpath/init-tools.log
+__PACKAGES_DIR=$__scriptpath/.nuget
+__TOOLRUNTIME_DIR=$__scriptpath/build_tools
+__DOTNET_PATH=$__scriptpath/.dotnet_stage0/$(uname -s)
```
Comment (eerhardt) [inline init-tools.sh]: I think we want to get rid of the OS name in the stage0 path.  Can we do it here?

File: init-tools.cmd (hunk 5 above)
Comment (eerhardt) [inline init-tools.cmd]: So we use the 32-bit CLI to build the 32-bit CLI.  The same reason we have stage1 and stage2 - test coverage.

File: build.sh (hunk 12 above)
Comment (brthor) [inline build.sh]: We should think about how to manage this, perhaps the build.sh script inside dotnet-cli-build should make the call to init-tools.sh so it's already in docker.

Comment (brthor) [inline build.sh]: long term I think we need some kind of outer shell build.sh and an inner build.sh that gets called in docker (not so different from what we have today).

Or we merge init-tools.sh with whatever the inner build.sh is, and just call that in docker, but then it would also need to invoke `dotnet build3`

File: init-tools.cmd (hunk 3 above)
Comment (sokket) [inline init-tools.cmd]: > This file is duplicated across repos, and trying to keep it "in sync" isn't going to be possible.

We aren't really keeping it in sync, rather we are following the design pattern that was put forth by corefx for initialize scripts

File: init-tools.cmd (hunk 5 above)
Comment (sokket) [inline init-tools.cmd]: But the CLI binary is managed code...why do we need the 32-bit CLI to build the 32-bit CLI?

File: init-tools.cmd (hunk 2 above)
Comment (sokket) [inline init-tools.cmd]: Fixed

File: init-tools.sh (hunk 9 above)
Comment (sokket) [inline init-tools.sh]: fixed

File: init-tools.sh (hunk 10 above)
Comment (sokket) [inline init-tools.sh]: fixed

File: build.cmd (hunk 11 above)
Comment (sokket) [inline build.cmd]: Removed

File: build.sh (hunk 12 above)
Comment (sokket) [inline build.sh]: Would this work:
```
dockerbuild()
{
    # Initialize the tooling and build
    $DIR/init-tools.sh "${args[@]}"
    BUILD_COMMAND=/opt/code/build_projects/dotnet-cli-build/build.sh $DIR/scripts/dockerrun.sh --non-interactive "$@"
}
```

File: init-tools.sh (hunk 13 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.sh (hunk 14 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: build.sh (hunk 12 above)
Comment (brthor) [inline build.sh]: No that won't work, the init tools is still run in the outer machine context.

`BUILD_COMMAND=/opt/code/build_projects/dotnet-cli-build/build.sh` Is the script run in the docker context.

Comment (sokket) [inline build.sh]: Ah and there isn't a way to run multiple commands?

Comment (brthor) [inline build.sh]: BUILD_COMMAND="/opt/code/init-tools.sh && /opt/code/build_projects/dotnet-cli-build/build.sh"

Comment (sokket) [inline build.sh]: Fixed

**Comment by sokket** [comment]
@dotnet-bot test this please

**Comment by sokket** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please

File: init-tools.cmd (hunk 3 above)
Comment (brthor) [inline init-tools.cmd]: I do agree that using powershell would be better fitting with designs that are currently in the cli repo. We've made changes throughout it's lifetime to use as little cmd as possible.

One reason for this is that the visual and conceptual differences between Bash and powershell is smaller than between bash and cmd. So when you need to maintain dual scripts it makes it that much easier to do so.

File: build.sh [hunk 15]
```diff
@@ -49,12 +49,15 @@ args=($temp)
 
 dockerbuild()
 {
-    BUILD_COMMAND=/opt/code/build_projects/dotnet-cli-build/build.sh $DIR/scripts/dockerrun.sh --non-interactive "$@"
+    # Initialize the tooling and build
+    BUILD_COMMAND="/opt/code/init-tools.sh && /opt/code/build_projects/dotnet-cli-build/build.sh $DIR/scripts/dockerrun.sh --non-interactive "$@""
```
Comment (brthor) [inline build.sh]: BuildCommand is not quoted correctly here.

`BUILD_COMMAND="/opt/code/init-tools.sh && /opt/code/build_projects/dotnet-cli-build/build.sh"`

It's setting an env var inside the dockerrun.sh script which then get's executed.

File: build_projects/dotnet-cli-build/build.ps1 [hunk 16]
```diff
@@ -46,28 +44,7 @@ cat "$RepoRoot\branchinfo.txt" | ForEach-Object {
...
-    
-& "$RepoRoot\scripts\obtain\dotnet-install.ps1" -Channel $env:CHANNEL -Architecture $Architecture -Verbose
-if($LASTEXITCODE -ne 0) { throw "Failed to install stage0" }
-
-# Put the stage0 on the path
-$env:PATH = "$env:DOTNET_INSTALL_DIR;$env:PATH"
-
-# Disable first run since we want to control all package sources
-$env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+$env:path = "$RepoRoot\.dotnet_stage0;" + $env:path
```
Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: I don't know if it matters on windows but should `path` be uppercase?

File: build_projects/dotnet-cli-build/build.ps1 [hunk 17]
```diff
@@ -46,28 +44,7 @@ cat "$RepoRoot\branchinfo.txt" | ForEach-Object {
...
-Write-Host "Installing .NET Core CLI Stage 0 from branchinfo channel"
-    
-& "$RepoRoot\scripts\obtain\dotnet-install.ps1" -Channel $env:CHANNEL -Architecture $Architecture -Verbose
-if($LASTEXITCODE -ne 0) { throw "Failed to install stage0" }
-
-# Put the stage0 on the path
-$env:PATH = "$env:DOTNET_INSTALL_DIR;$env:PATH"
-
-# Disable first run since we want to control all package sources
-$env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
```
Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: why did you remove this, but not in the `build_projects/dotnet-cli-build/build.sh `

File: init-tools.sh [hunk 18]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+__scriptpath="$(cd "$(dirname "$0")"; pwd -P)"
```
Comment (brthor) [inline init-tools.sh]: In other scripts in the repo we tend to name this DIR and use the full version of this which will resolve through symlinks.

https://github.com/dotnet/cli/blob/rel/1.0.0/scripts/dockerrun.sh#L15

File: init-tools.sh [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+then
+    echo "Increasing file description limit to 1024"
+    ulimit -n 1024
+fi
+
+# Disable first run since we want to control all package sources
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+if [ ! -e "$__PROJECT_JSON_FILE" ]; then
+    "$__scriptpath/scripts/obtain/dotnet-install.sh" --install-dir "$__DOTNET_PATH" --channel $__CHANNEL
```
Comment (brthor) [inline init-tools.sh]: Won't we want to get the latest dotnet stage0 even when buildtools is up to date?

File: build.sh (hunk 15 above)
Comment (sokket) [inline build.sh]: Fixed

File: init-tools.sh [hunk 20]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
```
Comment (brthor) [inline init-tools.sh]: we should use `set -e on this file so it will exit non-zero if any of the commands exit non-zero.

File: build_projects/dotnet-cli-build/build.ps1 (hunk 16 above)
Comment (sokket) [inline build_projects/dotnet-cli-build/build.ps1]: I copied this from another PS1 file in our branch...seems to work :)

File: init-tools.sh [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+    ulimit -n 1024
+fi
+
+# Disable first run since we want to control all package sources
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+if [ ! -e "$__PROJECT_JSON_FILE" ]; then
+    "$__scriptpath/scripts/obtain/dotnet-install.sh" --install-dir "$__DOTNET_PATH" --channel $__CHANNEL
+
+    if [ ! -d "$__PROJECT_JSON_PATH" ]; then mkdir -p "$__PROJECT_JSON_PATH"; fi
```
Comment (brthor) [inline init-tools.sh]: nit: mkdir -p can be called without this guard and it will not blow up even if the path exists.

File: init-tools.sh [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"
```
Comment (brthor) [inline init-tools.sh]: `echo "$__PROJECT_JSON_CONTENTS" > "$__PROJECT_JSON_FILE" `

File: init-tools.sh [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+        echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
+        echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        "$__DOTNET_CMD" restore "$__PROJECT_JSON_FILE" --packages "$__PACKAGES_DIR" --source "$__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details."; fi
+    fi
+
+    echo "Initializing BuildTools..."
+    echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    "$__BUILD_TOOLS_PATH/init-tools.sh" "$__scriptpath" "$__DOTNET_CMD" "$__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    if [ "$?" != "0" ]; then
```
Comment (brthor) [inline init-tools.sh]: If we use `set -e` we can remove these exit code checks.

File: init-tools.sh [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+# Disable first run since we want to control all package sources
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+if [ ! -e "$__PROJECT_JSON_FILE" ]; then
+    "$__scriptpath/scripts/obtain/dotnet-install.sh" --install-dir "$__DOTNET_PATH" --channel $__CHANNEL
+
+    if [ ! -d "$__PROJECT_JSON_PATH" ]; then mkdir -p "$__PROJECT_JSON_PATH"; fi
+    echo $__PROJECT_JSON_CONTENTS > "$__PROJECT_JSON_FILE"
+
+    if [ ! -e "$__BUILD_TOOLS_PATH" ]; then
```
Comment (brthor) [inline init-tools.sh]: is `$__BUILD_TOOLS_PATH` a directory? should the check be `-d`?

File: netci.groovy [hunk 25]
```diff
@@ -26,10 +26,10 @@ platformList.each { platform ->
 
     // Calculate the build command
     if (os == 'Windows_NT') {
-        buildCommand = ".\\build.cmd -Configuration ${configuration} -Architecture ${architecture} -Targets Default"
+        buildCommand = ".\\build.cmd -Configuration ${configuration} -Targets Default"
```
Comment (brthor) [inline netci.groovy]: I might have missed this discussion but how do we kick off x86 CI builds now?

File: build_projects/dotnet-cli-build/build.ps1 (hunk 17 above)
Comment (sokket) [inline build_projects/dotnet-cli-build/build.ps1]: Remove what specifically? :)

Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: > $env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1

https://github.com/dotnet/cli/pull/3719/files/174ce4c7d136c77a2cabf3b39175a149ff562c7e#diff-77fc6850a27234595d773b7c80f2280eL108

File: build_projects/dotnet-cli-build/build.ps1 (hunk 16 above)
Comment (brthor) [inline build_projects/dotnet-cli-build/build.ps1]: as long as it works :smile:

File: build_projects/dotnet-cli-build/build.ps1 (hunk 17 above)
Comment (sokket) [inline build_projects/dotnet-cli-build/build.ps1]: Oops, added back to Windows

File: init-tools.sh [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+        echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        "$__DOTNET_CMD" restore "$__PROJECT_JSON_FILE" --packages "$__PACKAGES_DIR" --source "$__BUILDTOOLS_SOURCE" >> "$__init_tools_log"
+        if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details."; fi
+    fi
+
+    echo "Initializing BuildTools..."
+    echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    "$__BUILD_TOOLS_PATH/init-tools.sh" "$__scriptpath" "$__DOTNET_CMD" "$__TOOLRUNTIME_DIR" >> "$__init_tools_log"
+    if [ "$?" != "0" ]; then
+        echo "ERROR: An error occured when trying to initialize the tools. Please check '$__init_tools_log' for more details."
```
Comment (brthor) [inline init-tools.sh]: Should StdErr be redirected as well in all of these above commands if we expect error details to be in the log file?

File: build_projects/update-dependencies/update-dependencies.ps1 [hunk 27]
```diff
@@ -17,20 +17,18 @@ if($Help)
...
-$Architecture='x64'
-
 $RepoRoot = "$PSScriptRoot\..\.."
 $AppPath = "$PSScriptRoot"
 
 # Use a repo-local install directory (but not the artifacts directory because that gets cleaned a lot
 if (!$env:DOTNET_INSTALL_DIR)
 {
-    $env:DOTNET_INSTALL_DIR="$RepoRoot\.dotnet_stage0\Windows\$Architecture"
+    $env:DOTNET_INSTALL_DIR="$RepoRoot\.dotnet_stage0\Windows"
```
Comment (eerhardt) [inline build_projects/update-dependencies/update-dependencies.ps1]: I think we want the "Windows" portion of this out of the path, but we want to keep the $Architecture piece.  That way we can build/test x64 and x86 in the same repo without cleaning in between.

Comment (eerhardt) [inline build_projects/update-dependencies/update-dependencies.ps1]: Also, the INSTALL dir between update-dependencies and init-tools should be the same place, so that way when we run them on our dev boxes, we don't download stage0 twice.

---
In reply to: [69064053](https://github.com/dotnet/cli/pull/3719#discussion_r69064053) [](ancestors = 69064053)

File: init-tools.cmd (hunk 5 above)
Comment (eerhardt) [inline init-tools.cmd]: Because we need do more than just build managed code in our repo.  For example, we crossgen managed code. Crossgen'ing needs a specific RID to work, i.e. we need the x86 RID.  The way our CrossGen tool figures out which RID to use is by looking at the architecture of the current process (i.e. the "build" process):

```C#
        private static string GetArch()
        {
#if NET451
            return Environment.Is64BitProcess ? "x64" : "x86";
#else
            return IntPtr.Size == 8 ? "x64" : "x86";
#endif
        }
```

And then later:

```
        private string GetCoreCLRRid()
        {
            string rid = null;
            if (CurrentPlatform.IsWindows)
            {
                var arch = RuntimeEnvironment.RuntimeArchitecture;
                rid = $"win7-{arch}";
            }
```

which is called by:

```
        private string GetCrossGenPackagePathForVersion()
        {
            string coreclrRid = GetCoreCLRRid();

            if (coreclrRid == null)
            {
                return null;
            }

            string packageId = $"runtime.{coreclrRid}.Microsoft.NETCore.Runtime.CoreCLR";

            return Path.Combine(
                Dirs.NuGetPackages,
                packageId,
                _coreClrVersion);
        }
```

Thus, the process needs to be x86 when we are building the x86 CLI.

File: init-tools.sh (hunk 18 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.cmd [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+if [%BUILDTOOLS_SOURCE%]==[] set BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
+set /P BUILDTOOLS_VERSION=< "%~dp0BuildToolsVersion.txt"
+set BUILD_TOOLS_PATH=%PACKAGES_DIR%Microsoft.DotNet.BuildTools\%BUILDTOOLS_VERSION%\lib\
+set PROJECT_JSON_PATH=%TOOLRUNTIME_DIR%\%BUILDTOOLS_VERSION%
+set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.json
+set PROJECT_JSON_CONTENTS={ "dependencies": { "Microsoft.DotNet.BuildTools": "%BUILDTOOLS_VERSION%" }, "frameworks": { "netcoreapp1.0": { } } }
+set BUILD_TOOLS_SEMAPHORE=%PROJECT_JSON_PATH%\init-tools.completed
+set DOTNET_INSTALL_ROOT_DIR=%~dp0.dotnet_stage0
+
+if DEFINED "%CHANNEL%" (
```
Comment (sokket) [inline init-tools.cmd]: Ugh the drawbacks of switching between OSX and Windows...one of my changes was dropped that defined this. Will add it back

File: init-tools.sh (hunk 20 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.sh (hunk 21 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.sh (hunk 22 above)
Comment (sokket) [inline init-tools.sh]: Why is that? The contents are a single line echo'd and the file is quoted...

File: init-tools.sh (hunk 23 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.sh (hunk 24 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: init-tools.sh (hunk 26 above)
Comment (sokket) [inline init-tools.sh]: Fixed

File: build_projects/update-dependencies/update-dependencies.ps1 (hunk 27 above)
Comment (sokket) [inline build_projects/update-dependencies/update-dependencies.ps1]: Why would we want to build two different architectures without cleaning in between? That leaves build artifacts and changes that should not be there and could have an adverse affect on the build...we should not leave any artifacts behind that could change the build in any way between builds

File: init-tools.sh (hunk 22 above)
Comment (brthor) [inline init-tools.sh]: bash is not the same as batch in that quotes are maintained when you output a variable. This can cause part of the string to be interpreted as arguments to echo if it's not quoted.

For example:
```
$ A="-e hello"
$ echo $A
hello
$ echo "$A"
-e hello
```

File: build_projects/update-dependencies/update-dependencies.ps1 (hunk 27 above)
Comment (eerhardt) [inline build_projects/update-dependencies/update-dependencies.ps1]: This is why our $(OutputDir) uses the $RID, so we can build for anything we want, and the build artifacts are separated.

File: init-tools.sh (hunk 22 above)
Comment (sokket) [inline init-tools.sh]: Good to know :) I'll add the quotes

File: build_projects/update-dependencies/update-dependencies.ps1 (hunk 27 above)
Comment (eerhardt) [inline build_projects/update-dependencies/update-dependencies.ps1]: The main point of the comment for this file is: 

"the INSTALL dir between update-dependencies and init-tools should be the same place, so that way when we run them on our dev boxes, we don't download stage0 twice."

Whever the "build" command puts Stage0, the "update-dependencies" command should also put Stage0.

File: init-tools.sh (hunk 19 above)
Comment (sokket) [inline init-tools.sh]: Do we want the latest stage0 every build? That seems excessive to me...

File: netci.groovy (hunk 25 above)
Comment (sokket) [inline netci.groovy]: I fixed the groovy script to execute x86 builds on an x86 console, which should be picked up correctly by the scripts and build the x86 versions

File: build_projects/update-dependencies/update-dependencies.ps1 (hunk 27 above)
Comment (sokket) [inline build_projects/update-dependencies/update-dependencies.ps1]: But it is more than just build artifacts...lock files are generated, output directories, tons of files and folders are created in numerous places outside of the artifacts folder that can cause changes to the build

File: init-tools.sh (hunk 19 above)
Comment (eerhardt) [inline init-tools.sh]: yes, it always checks if there is a new one.  no need to clean.  the check is fast.

File: netci.groovy (hunk 25 above)
Comment (brthor) [inline netci.groovy]: Let's ensure we can make the same change for VSO on mseng

Comment (sokket) [inline netci.groovy]: Should be a super simple change in VSO

File: build.cmd [hunk 29]
```diff
@@ -3,5 +3,8 @@
 REM Copyright (c) .NET Foundation and contributors. All rights reserved.
 REM Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+call %~dp0init-tools.cmd %*
+if %errorlevel% neq 0 exit /b %errorlevel%
```
Comment (livarcocc) [inline build.cmd]: With my change, you won't need this anymore. I added a init-tools.ps1 that build.ps1 is invoking already.

File: build.sh [hunk 30]
```diff
@@ -49,12 +49,15 @@ args=($temp)
 
 dockerbuild()
 {
-    BUILD_COMMAND=/opt/code/build_projects/dotnet-cli-build/build.sh $DIR/scripts/dockerrun.sh --non-interactive "$@"
+    # Initialize the tooling and build
+    BUILD_COMMAND="/opt/code/init-tools.sh && /opt/code/build_projects/dotnet-cli-build/build.sh" $DIR/scripts/dockerrun.sh --non-interactive "$@"
```
Comment (livarcocc) [inline build.sh]: You shoulnd't need this nor the change below either. init-tools is being invoked by build.sh already.

**Comment by sokket** [comment]
Closing in favor of #3811
